File: Helpers/NetworkManager.m
//
//  NetworkManager.m
//  MCompass
//
//  Created by D.Alex on 11/27/15.
//  Copyright © 2015 haizhi. All rights reserved.
//

#import "NetworkManager.h"

@implementation NetworkRequest

- (instancetype)initWithURL:(NSURL *)URL cachePolicy:(NSURLRequestCachePolicy)cachePolicy timeoutInterval:(NSTimeInterval)timeoutInterval startTime:(NSTimeInterval)startTime
{
    self = [super initWithURL:URL cachePolicy:cachePolicy timeoutInterval:timeoutInterval];
    if (self) {
        _startTime = startTime;
        _requestTimes = 0;
    }
    return self;
}
@end


@implementation NetworkManager {
    NSURLSession * _session;
}

+ (id)instance {
    static NetworkManager *sharedMyManager = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedMyManager = [[self alloc] init];
    });
    return sharedMyManager;
}

- (id)init {
    if (self = [super init]) {
        _session = [NSURLSession sharedSession];
    }
    return self;
}

- (void)dealloc {
    // Should never be called, but just here for clarity really.
}

/**
 网络请求fail逻辑:
 1. 或者网络不可用
 2. 或者请求时间达60秒
 3. 或者网络连接错误5次以上
 */
- (void) processRequest:(NetworkRequest *)request success: (void (^)(NSData * data))success fail: (void (^)(NSError * error))fail {
    NSURLSessionDataTask * task = [_session dataTaskWithRequest:request completionHandler:^(NSData * data, NSURLResponse * response, NSError * error) {
        if (error == nil) {
            success(data);
        } else {
            if (request.requestTimes++ >= 5) {
                fail(error);
            } else if ([[NSDate date] timeIntervalSince1970] >= request.startTime + 60) {
                fail(error);
            } else {
                [self processRequest:request success:success fail:fail]; //request self
            }
        }
    }];
    [task resume];
}

- (void) request:(NSString *)httpMethod url:(NSString *)url parameter:(NSData *)parameter success: (void (^)(NSData * data))success fail: (void (^)(NSError * error))fail {
    /*if ([AFNetworkReachabilityManager sharedManager].reachable != YES) {
     NSError * error = [[NSError alloc] initWithDomain:@"当前网络不可用" code:-1 userInfo:nil];
     fail(error);
     return;
     }*/
    
    if ([httpMethod isEqualToString:@"GET"]) {
        //string series parameter and append to url
    }
    
    NSURL * URL = [[NSURL alloc] initWithString:url];
    if (URL == nil) {
        NSError * error = [[NSError alloc] initWithDomain:@"error url" code:-1 userInfo:@{}];
        fail(error);
        return;
    }
    
    NetworkRequest * request =  [[NetworkRequest alloc] initWithURL:URL cachePolicy:NSURLRequestUseProtocolCachePolicy timeoutInterval:20 startTime:[[NSDate date] timeIntervalSince1970]];
    if ([httpMethod isEqualToString:@"POST"]) {
        request.HTTPMethod = httpMethod;
        request.HTTPBody = parameter;
        [request addValue:@"application/x-www-form-urlencoded" forHTTPHeaderField:@"Content-Type"];
    }
    
    [self processRequest:request success:success fail:fail];
}

- (void) POST:(NSString *)url parameter:(NSData *)parameter success: (void (^)(NSData * data))success fail: (void (^)(NSError * error))fail {
    [self request: @"POST" url:url parameter:parameter success:success fail:fail];
}

@end

File: Helpers/NetworkManager.swift
//
//  NetworkManager.swift
//  DEMO
//
//  Created by D.Alex on 12/21/15.
//  Copyright © 2015 Alex. All rights reserved.
//

import Foundation

class NetworkRequest: NSMutableURLRequest {
    var startTime:NSTimeInterval
    var retryTimes:Int
    
    override init(URL: NSURL, cachePolicy: NSURLRequestCachePolicy, timeoutInterval: NSTimeInterval) {
        self.startTime = NSTimeIntervalSince1970
        self.retryTimes = 0
        super.init(URL: URL, cachePolicy: cachePolicy, timeoutInterval: timeoutInterval)
    }
    
    convenience init(URL: NSURL, cachePolicy: NSURLRequestCachePolicy, timeoutInterval: NSTimeInterval, startTime:NSTimeInterval) {
        self.init(URL: URL, cachePolicy: cachePolicy, timeoutInterval: timeoutInterval)
        self.startTime = startTime
        
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

class NetworkManager: NSObject {
    
    static let instance = NetworkManager()
    var session:NSURLSession
    
    private override init() {
        session = NSURLSession.sharedSession()
        
        super.init()
        
    }
    deinit {
    }
    
    func generateRequest(httpMethod:String, url:String, parameter:NSData?) -> NetworkRequest? {
        
        guard let URL = NSURL(string: url) else {
            return nil
        }
        let request = NetworkRequest(URL: URL, cachePolicy: NSURLRequestCachePolicy.UseProtocolCachePolicy, timeoutInterval: 12, startTime: NSTimeIntervalSince1970)
        request.HTTPMethod = httpMethod
        if httpMethod == "POST" {
            request.HTTPBody = parameter
            request.addValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        }
        return request
    }
    
    func processTask(request:NetworkRequest, success:((data:NSData)->Void), fail:((error:NSError)->Void)) -> NSURLSessionDataTask {
        
        //check network available
        
        let task = self.session.dataTaskWithRequest(request) { (data, response, error) -> Void in
            if let data = data {
                success (data: data)
            } else {
                if request.retryTimes++ >= 5 || NSTimeIntervalSince1970 >= request.startTime + 60 {
                    if let error = error {
                        fail(error: error)
                    } else {
                        fail(error: NSError(domain: "requet timeout", code: -1, userInfo: nil))
                    }
                } else {
                    self.processTask(request, success: success, fail: fail)
                }
            }
        }
        task.resume()
        return task
    }
}

extension NetworkManager {
    
    //https://lvwenhan.com/ios/455.html
    
    class func POST(url:String, parameter:NSDictionary, success:((data: NSData) -> Void), fail:((error: NSError) -> Void)) -> Void {
        let parameter = NSKeyedArchiver.archivedDataWithRootObject(parameter)
        guard let request = NetworkManager.instance.generateRequest("POST", url: url, parameter: parameter) else {
            fail(error: NSError(domain: "generate request", code: -1, userInfo: ["url" : url, "parameter": parameter]))
            return
        }
        NetworkManager.instance.processTask(request, success: success, fail: fail)
    }
    
    class func GET(url:String, parameter:NSDictionary, success:((data: NSData) -> Void), fail:((error: NSError) -> Void)) {
        guard let request = NetworkManager.instance.generateRequest("GET", url: url, parameter: nil) else {
            fail(error: NSError(domain: "generate request", code: -1, userInfo: ["url" : url]))
            return
        }
        NetworkManager.instance.processTask(request, success: success, fail: fail)
    }
}



